My Site Dashboard: Tabs - Add "Site Menu" QS step, show QS focus point on tab

ashiagr opened the pull request:
Parent #16006

This PR adds `Site Menu` QS step when needed, displays QS focus point on the tab, and requests pending QS task on tab change. 
The "Site Menu" QS step is needed when QS task starts from `Dashboard`/ `Home` tab but the corresponding menu item is present on the `Site Menu` tab. In this case, an intermediate step is introduced that instructs users to tap `Site Menu` tab to continue.  

Video captures:

- "Site Menu" QS step (timstamp 0:23)
- Auto-scroll (on continue from "Site Menu" QS step (timstamp 0:26)
- FAB (timstamp 0:38)
- Integration from other parts of the app (timestamp 0.13 - Homepage, 0.31 - Publicize, 0.47 - Reader)

https://user-images.githubusercontent.com/1405144/159706819-c79d6d8e-2087-48ed-bf6b-ed679ad96a7c.mp4

To test:

**Test.1: Regression test Quick Start in single view with tabs disabled**

1. Launch the app.
2. Go to `My Site` -> `Me` -> `App Settings` -> `Debug settings`.
3. Turn off the `MySiteDashboardTabsFeatureConfig` flag under the `Features in development` section.
4. Restart the app.
5. Logout/login and select "Show me around" when the quick start prompt is shown.
6. Run tests in [wpandroid-quick-start-testing-instructions.txt](https://github.com/wordpress-mobile/WordPress-Android/files/8246096/wpandroid-quick-start-testing-instructions.txt).

**Test.2: Quick Start with tabs enabled**
1. Launch the app.
2. Go to `My Site` -> `Me` -> `App Settings` -> `Debug settings`.
3. Turn on the `MySiteDashboardTabsFeatureConfig` flag under the `Features in development` section.
4. Restart the app.
5. Logout/login and select "Show me around" when the quick start prompt is shown.
6. Note the "Quick Start" shows in the `Dashboard`/ `Home` tab only. 
7. Run tests in [wpandroid-quick-start-testing-instructions.txt](https://github.com/wordpress-mobile/WordPress-Android/files/8246096/wpandroid-quick-start-testing-instructions.txt).
[ Notice that in `Test 1 - Quick start notice display` in the text file, when quick start task is available only on `Site Menu` tab, `Site Menu` tab shows QS focus point. On clicking the tab, the focus point should disappear and the appropriate site menu item should get QS focus point.]

**Review Instructions**

Code review from only one reviewer is sufficient.

## Regression Notes
1. Potential unintended areas of impact
Quick start actions/tasks don't work as expected

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual and automated tested

3. What automated tests I added (or what prevented me from doing so)
Unit tests were updated to reflect refactoring efforts

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Register view pager page changed callback

ashiagr committed:
Request QS "Site Menu" step when needed

ashiagr committed:
Update tab focus point on QS Site Menu step request

ashiagr committed:
Clear QS "Site Menu" step and request pending task on tab change

ashiagr committed:
Tests: Adjust existing tests to compile

ashiagr committed:
Tests: Add block comments

ashiagr committed:
Tests: (Refactor) Optimize imports

ashiagr committed:
Tests: Add repository tests for quick start site menu step

ashiagr committed:
Tests: Add vm tests for quick start site menu step

ashiagr committed:
Refactor: Update tabs ui state conditionally

ashiagr committed:
Replace mediator with mutable live data

tiagomar commented:
👋  @ashiagr, I have only found the following issues:

<details>
  <summary>1. `Site Menu` focus point is not dismissed if another QS Item is selected</summary>

| Evidence | Steps | 
| - | - |
| <video width="300" src="https://user-images.githubusercontent.com/42008628/159987034-c77c9f11-7006-4f90-8192-a21215354596.mp4"> | Exploring QS Items in different order <br />1. `Home` -> `Customize your site` -> `Visit your site`<br />2. `Customize your site` -> `Visit your site` -> `Edit your homepage`. <br /> ❌ _Notice that you'll see two focus points._ <br />3. `Grow your audience` -> `Publish a post`.<br /> ❌ _Notice that you'll see two focus points._ |

</details>

<details>
  <summary>2. `View Site` focus point is not always displayed in Landscape mode</summary>

| Evidence | Steps | 
| - | - |
| <video width="300" src="https://user-images.githubusercontent.com/42008628/159993436-fd78c846-b6de-4097-bd63-a2d1375d0c13.mp4"> | 1. Portrait mode <br />2. `Home` -> `Customize your site` -> `Visit your site`<br />3. `Site Menu` <br /> ✅ _`View Site` focus point is displayed._  <br />4. Lanscape mode <br />5. `Home` -> `Customize your site` -> `Visit your site`<br />3. `Site Menu` <br /> ❌  _`View Site` focus point is displayed._  | 

</details>

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt:
```diff
@@ -140,7 +142,18 @@ class MySiteViewModel @Inject constructor(
+                /* We want to filter out tabs state livedata update when state is not set in uiModel.
+                   Without this check, tabs state livedata merge with state livedata may return a null state
+                   when building UiModel. */
+                uiModel.value?.state?.tabsUiState?.let {
```
Thank you for this comment ❤️

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/quickstart/QuickStartRepository.kt:
```diff
@@ -80,16 +80,17 @@ class QuickStartRepository
-    } else {
-        QuickStartOrigin.ALL
-    }
+    val quickStartOrigin = if (isMySiteTabsEnabled) QuickStartOrigin.DASHBOARD else QuickStartOrigin.ALL
```
Thanks for removing the "dashboard" here - I was also beginning to feel that it was redundant. 👏

zwarm commented on WordPress/src/test/java/org/wordpress/android/ui/mysite/cards/quickstart/QuickStartRepositoryTest.kt:
```diff
@@ -60,8 +61,17 @@ class QuickStartRepositoryTest : BaseUnitTest() {
+    private lateinit var quickStartSiteMenuStep: MutableList<QuickStartSiteMenuStep?>
     private val siteLocalId = 1
 
+    private val siteMenuTasks = listOf(
```
❓ for the future: How would this work if we show the qs card on the site_menu and not the dashboard?

ashiagr replied:
Actually, `isSiteMenuStepRequiredForTask` checks two conditions:

`quickStartTaskOrigin == MySiteTabType.DASHBOARD && task.showInSiteMenu()`

It doesn't matter which tab we show qs card, what matters is the quick start task origin which gets set on tab change.  Even if the qs card is displayed on the "site menu" tab, if we switch to the "dashboard" tab when the qs prompt is displayed, then we display the "site menu step". Hope it clarifies. 🤞

zwarm reviewed (approved):
Looking good @ashiagr . I left a few inline comments about good stuff.  I haven't had the chance to pull the branch down to walk through the new stuff in detail, but I did get to the tests.  I'm going to approve this, as nothing seems out of place, and I will do my own deeper dive tomorrow. I'll leave it up to you to decide if you want to merge it or wait.

Test 1 works as expected (single view, no tabs)
Test 2 works as expected. When on the dashboard tab, selecting an item that is only on "site menu", the notification dot is shown on the tab and then another notification dot when you are on the site menu. The notification dot is not shown again if the user navigates back to dashboard before completing the QS task. Also once the user moves over to the "site-menu" tab, all remaining notification dots show on that tab as expected.

zwarm commented:
👋 @tiagomar -  thanks for these two scenario's, I'll add them to future testing instructions. 
@ashiagr My findings are below:

> 1. `Site Menu` focus point is not dismissed if another QS Item is selected

It does get dismissed when you complete the "second" QS item. My uninformed guess is to clear the `_onQuickStartSiteMenuStep.postValue(null)` in `setActiveTask`. I tried it quickly, since I was running the tests to complete my review. :) 

> 2. `View Site` focus point is not always displayed in Landscape mode

I wasn't able to reproduce this one. I do see that when rotating the device, the focus points stay as expected.

ashiagr committed:
Clear site menu step on set active task

ashiagr committed:
Increase scroll delay to show QS focus point for site menu step pending task

In landscape mode, with smaller screen height, it takes longer to scroll to the site menu pending task item.

ashiagr commented:
Thanks, @tiagomar 🙏 

Both were valid issues. I've fixed them in commits:

1. `Site Menu` focus point is not dismissed if another QS Item is selected - 8d8bdfe
2. `View Site` focus point is not always displayed in Landscape mode - c3fe98e

Ready for another look. 🙇‍♀️

zwarm commented:
@ashiagr 👍  I reviewed the changes and all works now. Thank you

ashiagr commented:
Thanks, @zwarm. @tiagomar I'm merging this PR as we need it for the next one. Feel free to share any feedback and I'll address it separately.

tiagomar commented:
@ashiagr @zwarm  The changes look good! 

I ran into an issue while testing that it is not related to this card but I thought it was good to mention it here to give visibility and maybe get some help finding the best way to address it. 

Focus point is not displayed after rotating the phone in steps list: #16190
